### PR TITLE
Introduce buffered audit backend

### LINKED ISSUE #53006: Introduce buffer to audit log output
We should introduce a buffer (similar to what we have in the webhook output backend) to the log file output backend. Audit events should be queued on a relatively large buffer chanel, and an output go routine should grab events off the queue and write them to disk. If an audit event is output when the channel is full, it should be dropped (this should be counted by a prometheus metric, and an e2e test should ensure that the count is below some threshold). This behavior should be optional (configurable)

/cc @crassirostris 

Mitigates: https://github.com/kubernetes/kubernetes/issues/51899

### PR DESCRIPTION
**Pull request by crassirostris**
This is a copy from https://github.com/kubernetes/kubernetes/issues/53006 with comments addressed, since @hzxuzhonghu cannot work on this right now and the code freeze is close. Thanks @hzxuzhonghu for the contribution!

This is the first PR to fix https://github.com/kubernetes/kubernetes/issues/53006 and then https://github.com/kubernetes/kubernetes/issues/53020

This PR introduces a new audit backend, that batches incoming audit events and sends these batches to the underlying audit backend. This new backend is designed to replace existing batching webhook backend and allow to use batching with log backend, to solve the performance issues.

@sttts @ericchiang @CaoShuFeng @tallclair would really appreciate your review

```release-note
Buffered audit backend is introduced, to be used with other audit backends.
```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/plugin/pkg/audit/buffered/buffered.go [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+	// This WaitGroup is used to wait for all sending routines to finish before shutting down audit backend.
+	wg sync.WaitGroup
+
+	// Limits the number of batches sent to the delegate backend per second.
+	throttle flowcontrol.RateLimiter
+}
+
+var _ audit.Backend = &bufferedBackend{}
+
+// NewBackend returns a buffered audit backend wraps delegate backend.
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/plugin/pkg/audit/buffered/buffered.go]: nit: that wraps the delegate backend.

File: staging/src/k8s.io/apiserver/plugin/pkg/audit/buffered/buffered.go [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+func NewBackend(backend audit.Backend, config BatchConfig) audit.Backend {
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/plugin/pkg/audit/buffered/buffered.go]: s/backend/delegate/

**Comment by sttts** [comment]
Looks good so far. Missing a commit to integrate it, don't I?

**Comment by crassirostris** [comment]
@sttts 

> Missing a commit to integrate it, don't I?

I'll do it in a separate PR, to make it easier to review

**Comment by sttts** [comment]
> I'll do it in a separate PR, to make it easier to review

Would help to see the removal of the old buffering code here. I guess all the logic is just copied and refactored to be generic?

**Comment by crassirostris** [comment]
@sttts Yes, but removing the old code requires changing the code related to parsing command-line parameters, that's why I wanted to move it to a separate PR

**Comment by sttts** [comment]
@crassirostris can you open that second PR?

**Comment by crassirostris** [comment]
@sttts OK, I found a couple of nits related to the integration process, agree it's easier to make it all in one PR

Will update it soon

**Comment by sttts** [comment]
> agree it's easier to make it all in one PR

two PRs are fine. But seeing the WIP second PR helps :)

**Comment by crassirostris** [comment]
@sttts PTAL, I created a PR, based on this one: https://github.com/kubernetes/kubernetes/pull/60237

**Review by tallclair — COMMENTED** [review]
/lgtm

**Comment by sttts** [comment]
Left some non-critical comments on https://github.com/kubernetes/kubernetes/pull/60237.

**Comment by sttts** [comment]
/approve

Removed the release note as this is not user facing (yet). For the record, it was: "Audit buffered backend is introduced, to be used with any other backend to add batching and make processing audit events asynchronous."

**Pull request merged**